### ISSUE TITLE
Follow is "Awaiting approval" on Mastodon

### ISSUE THREAD

**Issue by noplanman** [jdormit/pterotype#15]
(**Edit**: Forgot to add earlier)
- WP 5.0
- Pterotype 1.2.13
- Blog: https://freeflowingfusion.com

When following my blog on Mastodon, it's awaiting approval...

How and where can I approve the follow?
And why is this happening in the first place?

@jdormit I noticed on your demo instances that it has the same webfinger data, so not sure what's up here.

**Comment by scottandrewlepera** [comment]
Can confirm. I can see the Follow request (and subsequent Unfollows) in the inbox JSON, but my website doesn't seem to auto-approve the request, even though I can see the `manuallyApprovesFollowers` item is set to false.

Using WP 4.9

(Update) I also turned on WP debugging while performing the follow/unfollow actions from Mastodon, and saw nothing in the debug.log, so I see no evidence of any errors reported during those actions.

**Comment by noplanman** [comment]
@jdormit Any other info I can provide to figure this out? Or is it something super-obvious I'm missing? :confused:

**Comment by jdormit** [comment]
I've seen this reported before from other beta testers. Honestly, I haven't been able to find the time to dig into what's happening yet. Here's what's supposed to happen: Pterotype receives the follow request and automatically sends back an Accept activity a few seconds later. On top of that, Pterotype sets `manuallyApprovesFollowers` to `false` in the Actor definitions, which should prevent Mastodon from displaying the "waiting for approval" icon at all.

My next steps when I find the time will be to send follow requests to your instances manually and see if I get the Accept response back, which will at least let me narrow down what could be going wrong.

Full disclosure, follow/accept are part of the feature set that is going to be rewritten for 2.0, so if I don't seem to be treating these reports as high priority as I should that's why.

**Comment by noplanman** [comment]
@jdormit Ok, cool! Looking forward to the updates :innocent: 

Not very urgent for me just yet, as I'm mainly exploring new frontiers at the moment, to pitch new possibilities to my clients.

**Comment by carlydotwebsite** [comment]
Hi! I was working on on a similar Wordpress plugin when I heard about yours, and I've had similar issues getting Mastodon to report the follow activity as accepted. As far as I can tell from combing through the Mastodon source code, manuallyApprovesFollowers doesn't necessarily do anything if it's on a remote authority (e.g. what we're doing).

I was eventually able to get the follow to complete by sending back an accept request after doing a lot of trial and error with the headers in Postman. The relevant snippet starts here (sorry it's kind of a mess), if you're interested!
https://github.com/cmho/wp-to-activitypub/blob/master/activitypub.php#L573

**Comment by jdormit** [comment]
Hey @cmho, thanks! Great to see other devs bridging the WordPress-Fediverse gap. Pterotype does currently send back an Accept in response to receiving a Follow - seems like something is going wrong with that process for certain folks.

@noplanman @scottandrewlepera can you give me more details about your hosting setup? What web server are you using, etc.? I ask because the mechanism by which Pterotype ensures that the Accept gets back to the requesting server after the Follow has already been processed is to process the Follow asynchronously by sending a request back to its own server (a dirty hack because PHP is single-threaded). I'm wondering if your setup is preventing that self-request from going through.

**Comment by carlydotwebsite** [comment]
@jdormit It might also be that you need to send Host as a header in addition to in the signature? This is just a guess, but Mastodon seems to be kind of picky about what headers it requires in order to process activities.

**Comment by noplanman** [comment]
@jdormit I have my WordPress running on [Uberspace.de](https://uberspace.de/), a shared hosting.

It's running in a WP Multisite setup, maybe that's the issue?

Before I go exploring too deeply, could you point me to your code that does the neat hack? Then I can try to execute that in a more debug-like fashion to see if I can see who's not playing nicely :smile:

**Comment by DeadSuperHero** [comment]
Just wanted to follow up on this: my Mastodon instance can follow Pterotype blogs flawlessly, but my Pleroma instance is perpetually stuck on the "Awaiting Approval" sequence. I'm not really sure how to clear that hurdle.

**Comment by noplanman** [comment]
@DeadSuperHero So you're saying it's a Mastodon instance issue?
Can you follow `@blog@freeflowingfusion.com` without a problem too?

Just wondering, if that's the case, how to get my Mastodon to play nice :thinking:

**Comment by DeadSuperHero** [comment]
@noplanman no, it works fine with Mastodon. I'm just stuck with Pleroma, which only advances to `Follow Requested`. I just tested with your blog, same effects as mine.

I wonder if Pleroma is searching for something missing from my nginx configuration? I set up NextCloud recently, and it required me to expose some stuff for webfinger discovery...

**Comment by jdormit** [comment]
I can look into more details later, but this is a know problem with
Pterotype right now. I'm not sure what Pleroma is looking for that we are
not providing though...

**Comment by noplanman** [comment]
> I just tested with your blog, same effects as mine.

@DeadSuperHero So the follow worked correctly then?
I have the same `Awaiting approval` with the Nexcloud Social addon :confused: 
So (as an off-topic) I'd be interested in a fix for that, if and when you figure out what's up.

**Comment by jdormit** [comment]
Hmm, so it sounds like Pterotype probably isn't sending the Accept activity
correctly, but Mastodon is still accepting it for some reason. The next
step is probably running Pleroma locally and debugging to see what it
expects to happen.

On Dec 31, 2018 5:54 PM, "Armando Lüscher" <notifications@github.com> wrote:

I just tested with your blog, same effects as mine.

@DeadSuperHero <https://github.com/DeadSuperHero> So the follow worked
correctly then?
I have the same Awaiting approval with the Nexcloud Social addon 😕

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/pterotype-project/pterotype/issues/15#issuecomment-450697153>,
or mute the thread

**Comment by DeadSuperHero** [comment]
Hey @jdormit, is there any way to produce output between my Wordpress and Pleroma instance?
If there's something helpful that I can do to help move the dial here, I'd be happy.to help!

**Comment by jdormit** [comment]
Do you run the Pleroma instance? It'd be super helpful to try to follow your WordPress blog from your Pleroma instance and check the Pleroma logs/database to see what actually happened. Here's what should happen:

1. Pleroma sends a Follow activity to Pterotype
2. Pterotype saves the Follow activity, then waits 5 seconds and sends back an Accept activity
3. Pleroma receives the Accept activity, saves it, and updates your Pleroma actor to be following the Pterotype actor

So something is either going wrong with Pterotype sending back the Accept or with Pleroma receiving it. It'd be super valuable to know which one. My instinct is that Pleroma doesn't like the Accept activity it is getting sent for some reason - maybe it doesn't like the HTTP signature? - but it could be something else entirely.

**Comment by mcg** [comment]
Since these are likely related pfefferle/wordpress-activitypub#19 and pixelfed/pixelfed#664

**Comment by jdormit** [comment]
Yeah seems like it. I've been putting this off since I'm in the process of rewriting all of the Pterotype ActivityPub code - I'll revisit when V2 is released.

See https://github.com/pterotype-project/activitypub-php/issues/5.

**Comment by Cambridgeport90** [comment]
Awaiting V.@.0.
